File: java/scf-core/src/main/java/org/mydotey/scf/DefaultPropertyChangeEvent.java
package org.mydotey.scf;

/**
 * @author koqizhao
 *
 * Jul 19, 2018
 */
public class DefaultPropertyChangeEvent<K, V> implements PropertyChangeEvent<K, V> {

    private Property<K, V> _property;
    private V _oldValue;
    private V _newValue;
    private long _changeTime;

    public DefaultPropertyChangeEvent(Property<K, V> property, V oldValue, V newValue) {
        this(property, oldValue, newValue, System.currentTimeMillis());
    }

    public DefaultPropertyChangeEvent(Property<K, V> property, V oldValue, V newValue, long changeTime) {
        _property = property;
        _oldValue = oldValue;
        _newValue = newValue;
        _changeTime = changeTime;
    }

    @Override
    public Property<K, V> getProperty() {
        return _property;
    }

    @Override
    public V getOldValue() {
        return _oldValue;
    }

    @Override
    public V getNewValue() {
        return _newValue;
    }

    @Override
    public long getChangeTime() {
        return _changeTime;
    }

    @Override
    public String toString() {
        return String.format("%s { property: %s, oldValue: %s, newValue: %s, changeTime: %s }",
                getClass().getSimpleName(), _property, _oldValue, _newValue, _changeTime);
    }

}

File: dotnet/src/MyDotey.SCF/DefaultPropertyChangeEvent.cs
using System;

namespace MyDotey.SCF
{
    /**
     * @author koqizhao
     *
     * Jul 19, 2018
     */
    public class DefaultPropertyChangeEvent<K, V> : IPropertyChangeEvent<K, V>
    {
        public virtual IProperty<K, V> Property { get; private set; }
        public virtual V OldValue { get; private set; }
        public virtual DateTime ChangeTime { get; private set; }
        public virtual V NewValue { get; private set; }

        public DefaultPropertyChangeEvent(IProperty<K, V> property, V oldValue, V newValue)
            : this(property, oldValue, newValue, DateTime.Now)
        {

        }

        public DefaultPropertyChangeEvent(IProperty<K, V> property, V oldValue, V newValue, DateTime changeTime)
        {
            if (property == null)
                throw new ArgumentNullException("property is null");

            Property = property;
            OldValue = oldValue;
            NewValue = newValue;
            ChangeTime = changeTime;
        }

        IProperty IPropertyChangeEvent.Property { get { return Property; } }

        object IPropertyChangeEvent.OldValue { get { return OldValue; } }

        object IPropertyChangeEvent.NewValue { get { return NewValue; } }

        public override String ToString()
        {
            return string.Format("{0} {{ property: {1}, oldValue: {2}, newValue: {3}, changeTime: {4} }}",
                    GetType().Name, Property, OldValue, NewValue, ChangeTime);
        }
    }
}